lib/Dialect/FIRRTL/Transforms/RemoveUnusedPorts.cpp

```diff
@@ -0,0 +1,183 @@
+        if (!connect || !isa_and_nonnull<InvalidValueOp, ConstantOp>(
+                            connect.src().getDefiningOp()))
+          continue;
+
+        Operation *srcOp = connect.src().getDefiningOp();
+        if (auto constant = dyn_cast<ConstantOp>(srcOp))
+          outputPortConstants.push_back(constant.value());
+        else {
+          assert(isa<InvalidValueOp>(srcOp) && "only expect invalid");
+          outputPortConstants.push_back(None);
+        }
+
+        // Erase connect op because we are going to remove this output ports.
+        connect.erase();
+
+        if (srcOp->use_empty())
+          srcOp->erase();
+      } else {
+        // Otherwise, we cannot remove the port.
+        continue;
+      }
+    }
+
+    removalPortIndexes.push_back(index);
+  }
+
+  // If there is nothing to remove, abort.
+  if (removalPortIndexes.empty())
+    return;
+
+  // Delete ports from the module.
+  module.erasePorts(removalPortIndexes);
+  LLVM_DEBUG(llvm::for_each(removalPortIndexes, [&](unsigned index) {
+               llvm::dbgs() << "Delete port: " << ports[index].name << "\n";
+             }););
+
+  // Rewrite all uses.
+  for (auto *use : instanceGraphNode->uses()) {
+    auto instance = use->getInstance();
+    ImplicitLocOpBuilder builder(instance.getLoc(), instance);
+    unsigned outputPortIndex = 0;
+    for (auto index : removalPortIndexes) {
+      auto result = instance.getResult(index);
+      assert(!ports[index].isInOut() && "don't expect inout ports");
+
+      // If the port is input, replace the port with an unwritten wire
+      // so that we can remove use-chains in SV dialect canonicalization.
+      if (ports[index].isInput()) {
+        WireOp wire = builder.create<WireOp>(result.getType());
+
+        // Check that the input port is only written. Sometimes input ports are
+        // used as temporary wires. In that case, we cannot erase connections.
+        bool onlyWritten = llvm::all_of(result.getUsers(), [&](Operation *op) {
+          if (auto connect = dyn_cast<ConnectOp>(op))
+            return connect.dest() == result;
+          return false;
+        });
+
+        result.replaceUsesWithIf(wire, [&](OpOperand &op) -> bool {
+          // Connects can be deleted directly.
+          if (onlyWritten && isa<ConnectOp>(op.getOwner())) {
+            op.getOwner()->erase();
+            return false;
+          }
+          return true;
+        });
+
+        // If the wire doesn't have an user, just erase it.
+        if (wire.use_empty())
+          wire.erase();
+
+        continue;
+      }
+
+      // Output port.  Replace with the output port with an invalid or constan
+      // value.
+      auto portConstant = outputPortConstants[outputPortIndex++];
+      Value value;
+      if (portConstant)
+        value = builder.create<ConstantOp>(*portConstant);
+      else
+        value = builder.create<InvalidValueOp>(result.getType());
+
+      result.replaceAllUsesWith(value);
```

Comment:
I'm not sure there is a case output ports being used as sink values, is that possible?

Reply:
The case I'm concerned about is something where the only usage of an output port is as a RHS value. (An output port being used as a source value.)

```
module Bar:
  output a: UInt<1>
  output b: UInt<1>

  a is invalid
  b <= a
```

Reply:
Sorry my explanation was not enough. I was considering the caller side.  

```
module Bar:
  output a: UInt<1>

  a is invalid
module Foo:
    output b : UInt<1>
    input c : UInt<1>
  inst bar of Bar
  b <= bar.a
  bar.a <= c
```

But anyway, it raises a verification error so it should be fine. Thanks!

(Btw, note that for your example above, neither a nor b is erased only with this pass. 
I'm intending running this pass after IMCP.)

Reply:
Sorry, a better example would be this:

```scala
circuit Foo:
  module Bar:
    input a: UInt<1>
    output b: UInt<1>
    output c: UInt<1>

    b <= a
    c <= b

  module Foo:
    input a: UInt<1>
    output c: UInt<1>

    inst bar of Bar
    bar.a <= a
    c <= bar.c
```

Output `Bar.b` is unused from the outside, but is used on the inside. The pass should still do the right thing on this circuit even without IMCP.

Reply:
Thanks, this example makes sense. Currently this pass can not remove Bar.b but I think I can add this functionality without much difficulty.

Reply:
SFC is _not_ removing it and it's good that this pass does the same thing. (If IMCP or something makes `Bar.b` unused by optimizing `c <= b` -> `c <= a` then I think it's fine to remove it. However, this pass shouldn't be aggressively searching for evidence to remove things.)

Reply:
Yeah, I also don't want to do data-flow analysis aggressively within this pass. 

But I think it is sufficient to check `Bar.b` is not read in any caller. We can replace `b` with a local wire with in Bar so that the created wire will be probably eliminated by sv wire canonicalizer: 
https://github.com/llvm/circt/blob/7205a2cb2dd5c30d5663574e867f0111ed105a47/lib/Dialect/SV/SVOps.cpp#L1022-L1023